fdbserver/MoveKeys.actor.cpp

```diff
@@ -2668,12 +3025,47 @@ ACTOR Future<Void> cleanUpDataMoveCore(Database occ,
 				// For each intersecting range, clear existing dest servers and checkpoints on all src servers.
 				state int i = 0;
 				for (; i < currentShards.size() - 1; ++i) {
-					KeyRangeRef rangeIntersectKeys(currentShards[i].key, currentShards[i + 1].key);
-					std::vector<UID> src;
-					std::vector<UID> dest;
-					UID srcId, destId;
+					state KeyRangeRef rangeIntersectKeys(currentShards[i].key, currentShards[i + 1].key);
+					state std::vector<UID> src;
+					state std::vector<UID> dest;
+					state UID srcId;
+					state UID destId;
 					decodeKeyServersValue(UIDtoTagMap, currentShards[i].value, src, dest, srcId, destId);
 
+					// Pre validate consistency of update of keyServers and serverKeys
+					if (SERVER_KNOBS->AUDIT_DATAMOVE_PRE_CHECK &&
+					    auditStorageFailedCount < SERVER_KNOBS->AUDIT_DATAMOVE_MAX_SUCCESSIVE_FAILED) {
+						state std::vector<Future<Void>> validationActors;
+						state std::unordered_map<UID, bool> validationResult;
+						for (auto& ssid : src) {
+							validationActors.push_back(
+							    store(validationResult[ssid],
+							          validateRangeAssignment(
+							              &tr, rangeIntersectKeys, ssid, "cleanUpDataMoveCore_precheck_src")));
+						}
+						for (auto& ssid : dest) {
+							validationActors.push_back(
+							    store(validationResult[ssid],
+							          validateRangeAssignment(
+							              &tr, rangeIntersectKeys, ssid, "cleanUpDataMoveCore_precheck_dest")));
+						}
+						wait(waitForAll(validationActors));
+						bool anyErrorDetected = false;
+						for (const auto& [ssid, consistent] : validationResult) {
+							if (!consistent) {
+								anyErrorDetected = true;
+								TraceEvent(SevError, "AuditStorageError")
+								    .detail("Context", "cleanUpDataMoveCore_precheck")
+								    .detail("WrongSS", ssid);
+							}
+						}
+						if (anyErrorDetected) {
+							int _ = wait(setDDMode(occ, 0));
+							throw audit_storage_error();
+						}
+						auditStorageFailedCount = 0;
+					}
```

Comment:
This code block is a common pattern, can we put it into a `ACTOR`, so that it can be shared instead of c/p the code block.

Reply:
Yes.